File: src/api/types/agent.ts
import type { DiffStats } from "./git";

export type AgentStatus = "spawning" | "running" | "idle" | "stopped" | "error";

export type AgentView = "custom" | "native";

/** What a forked workspace's worktree starts from. `clean` forks the parent's
 *  base branch; `carry` overlays the parent's current working tree (incl.
 *  uncommitted work) so the fork builds on unmerged work. Mirrors the backend
 *  `ForkCode`. */
export type ForkCode = "clean" | "carry";

/** How much of the parent conversation a fork carries. Mirrors the backend
 *  `ForkContext`. Summarized context ships in a follow-up slice. */
export type ForkContext =
  | { kind: "none" }
  | { kind: "full" }
  | { kind: "up_to_message"; prompt: number };

export interface TrackedRepo {
  repo_path: string;
  subdir: string;
  branch?: string | null;
  parent_branch?: string | null;
  /** Bound PR identity + last persisted snapshot (see prSnapshot in
   *  util/prState.ts). Written by the backend on every successful PR fetch;
   *  what the UI renders when GitHub or the checkout is unavailable. */
  pr_number?: number | null;
  pr_url?: string | null;
  pr_title?: string | null;
  pr_state?: string | null;
  /** The repo's display label within its project ("Frontend"); null falls
   *  back to the folder basename. Denormalized from the repos table. */
  label?: string | null;
}

export interface ArchivedRepoSnapshot {
  repo_path: string;
  subdir: string;
  branch_name?: string | null;
  branch_tip_sha?: string | null;
  parent_branch?: string | null;
  parent_branch_sha?: string | null;
  diff_stats: DiffStats;
}

export interface ArchiveMetadata {
  archived_at: string;
  repos: ArchivedRepoSnapshot[];
  diff_stats: DiffStats;
}

export interface AgentRecord {
  id: string;
  /** Project this agent belongs to. Used to scope project-level
   *  settings (e.g. Run panel overrides). */
  project_id: string;
  name: string;
  /** Which CLI backend powers this agent (claude, codex, ...). Maps to
   *  the TS adapter registered under the same id. */
  provider: string;
  /** Repos this agent has checkouts in. Always non-empty;
   *  `repos[0]` is the primary (the workspace repo at spawn). */
  repos: TrackedRepo[];
  task: string;
  status: AgentStatus;
  view: AgentView;
  session_id?: string | null;
  created_at: string;
  last_error?: string | null;
  /** Set when the agent has been archived. Live agents have null. */
  archive?: ArchiveMetadata | null;
  /** Claude's session-level reasoning effort (`--effort <level>`), chosen at
   *  spawn. Null for agents where effort wasn't set or isn't session-scoped. */
  effort?: string | null;
  /** Model chosen at spawn. Null/undefined means the provider CLI default. */
  model?: string | null;
  /** The custom agent this session was spawned from (null for a built-in
   *  spawn). Used to show the custom agent's name/color in the sidebar. */
  custom_agent_id?: string | null;
  /** Sandbox engine stamped at creation ("sandbox-exec" | "docker") and kept
   *  for the agent's life — a settings change never re-engines it. Null for
   *  agents created before engine selection existed (they run sandbox-exec). */
  sandbox_engine?: string | null;
  /** The GitHub issue this workspace was started from (bare issue number as
   *  text), set by the Home inbox's "Start work". Null for a normal spawn. */
  issue_ref?: string | null;
  /** What this workspace is for, when a surface other than the sidebar owns it
   *  (see [`ROADMAP_PM_PURPOSE`]). Null — the normal case — is a sidebar agent.
   *  Tagged workspaces are absent from the `get_workspace` snapshot entirely;
   *  they arrive through `listProjectChats`. */
  purpose?: string | null;
  /** A short title for the work, written by the agent via the `set_title`
   *  mailbox op once it knows what the user wants. Null until then; the
   *  sidebar subtitle falls back to the first line of `task`. */
  title?: string | null;
}

/** `workspaces.purpose` for a Roadmap project-manager chat — a manual chat that
 *  lives only on a project's Roadmap tab, never in the sidebar (which is for
 *  feature-development agents and workflow runs). Mirrors the backend's
 *  `workspace::PURPOSE_ROADMAP_PM`; the backend also denies these workspaces the
 *  publish ops, so a PM chat can read the code but never ship it. */
export const ROADMAP_PM_PURPOSE = "roadmap-pm";

/** A pinned repo joined with its owning project. `name` is the project's
 *  user-editable display name (defaults to the folder basename, survives
 *  rename/relocate); `project_id` addresses the project for rename/relocate +
 *  per-project settings. `label` names this repo within a multi-repo project
 *  ("Frontend", "Gateway"); null falls back to the folder basename. */
export interface ProjectRef {
  path: string;
  name: string;
  project_id: string;
  label: string | null;
}

export interface Workspace {
  /** Repos pinned in the sidebar. Empty on first launch. */
  repos: string[];
  /** Per-repo project metadata, parallel to `repos`. */
  projects: ProjectRef[];
  agents: AgentRecord[];
}

export interface ProjectDeleteResult {
  workspace: Workspace;
  deleted_agent_ids: string[];
  deleted_run_ids: string[];
}

export interface AgentOutputEvent {
  agent_id: string;
  /** Raw PTY bytes, base64-encoded over IPC. Decode with `decodeBase64`. */
  bytes: string;
}

/** Raw stream-json event from claude in custom view. Shape varies by
 *  `type`; the UI pattern-matches. */
export interface AgentManagedEvent {
  agent_id: string;
  event: Record<string, unknown> & { type?: string };
  /** The agent's per-event count under the current host process, matching
   *  `LiveTurn.next_seq`: a client that replayed the running turn skips frames
   *  numbered below the snapshot it folded. Absent from hosts before it. */
  seq?: number;
}

export interface AgentStatusEvent {
  agent_id: string;
  status: AgentStatus;
  last_error: string | null;
}

/** Which step of a fresh spawn is running, from `agent:spawn-progress`. A host
 *  newer than this client can name a stage that is not here, so readers must
 *  tolerate an unknown value rather than index blindly. */
export type SpawnStage =
  | "preparing"
  | "cloning"
  | "indexing"
  | "carrying"
  | "attaching_repos"
  | "starting";

/** Progress behind the `spawning` status — a label for the wait, never
 *  authoritative state. Not snapshotted: a client that connects mid-spawn has
 *  no stage until the next one fires. */
export interface AgentSpawnProgressEvent {
  agent_id: string;
  stage: SpawnStage;
  /** What the stage is working on, when it has something to name (the repo
   *  being checked out, say). */
  detail: string | null;
}

export interface AgentViewEvent {
  agent_id: string;
  view: AgentView;
}

export interface AgentEffortEvent {
  agent_id: string;
  effort: string | null;
}

export interface AgentModelEvent {
  agent_id: string;
  model: string | null;
}

export interface AgentTaskEvent {
  agent_id: string;
  task: string;
}

export interface AgentTitleEvent {
  agent_id: string;
  title: string;
}

export interface AgentBranchEvent {
  agent_id: string;
  subdir: string;
  branch: string;
}

export interface AgentRepoAddedEvent {
  agent_id: string;
  repo: TrackedRepo;
}

/** A successful mutating git op (`commit`/`push`/`PR`/`update`) the agent ran
 *  this turn — the ground-truth signal that a delegated git action happened. */
export interface AgentGitActionEvent {
  agent_id: string;
  op: string;
}

export interface ShellOutputEvent {
  agent_id: string;
  /** Raw PTY bytes, base64-encoded over IPC. Decode with `decodeBase64`. */
  bytes: string;
}

File: src-tauri/src/commands/agent.rs
//! Agent lifecycle: spawn / fork, input (write, message, tool-use answers),
//! terminal control, run-state transitions, and per-agent repo attach.

use std::path::PathBuf;
use std::sync::Arc;
use tauri::State;

use crate::error::Result;
use crate::host::EngineCtx;
use crate::managed_session::ToolUseBehavior;
use crate::supervisor::Supervisor;
use crate::workspace::{AgentRecord, AgentView, TrackedRepo};

// Args mirror the frontend `invoke("spawn_agent", ...)` payload one-to-one;
// they're the IPC wire surface, not collapsible into a struct here.
#[allow(clippy::too_many_arguments)]
#[tauri::command]
pub async fn spawn_agent(
    supervisor: State<'_, Arc<Supervisor>>,
    ctx: State<'_, Arc<EngineCtx>>,
    view: Option<AgentView>,
    repo_path: String,
    provider: Option<String>,
    name: Option<String>,
    effort: Option<String>,
    model: Option<String>,
    instructions: Option<String>,
    custom_agent_id: Option<String>,
    skills: Option<Vec<crate::agent_profile::SkillSnapshot>>,
    mcp_servers: Option<Vec<crate::agent_profile::McpServerSnapshot>>,
    fork_base: Option<String>,
    issue_ref: Option<String>,
    // Tags a workspace owned by a surface other than the sidebar — today only
    // the Roadmap tab's project-manager chat (`workspace::PURPOSE_ROADMAP_PM`).
    // Absent for a normal spawn.
    purpose: Option<String>,
    // The first prompt, persisted as the task at creation so the sidebar row
    // never shows an empty task while the process starts. Absent when the
    // caller has no prompt yet.
    task: Option<String>,
) -> Result<AgentRecord> {
    fletch_core::commands::spawn_agent_impl(
        supervisor.inner().clone(),
        ctx.inner().clone(),
        view,
        repo_path,
        provider,
        name,
        effort,
        model,
        instructions,
        custom_agent_id,
        skills,
        mcp_servers,
        fork_base,
        issue_ref,
        purpose,
        task,
    )
    .await
}

/// Fork an existing workspace into a new one, seeding its worktree (`code`) and
/// conversation (`context`) independently. `context = up_to_message` carries the
/// parent conversation through the navigable prompt at a 0-based ordinal (the
/// same ordinal the chat's turn list uses; git-action turns excluded).
///
/// `context_digest` is the frontend-rendered prose for the carried range — built
/// there so it renders uniformly across every provider's chat adapter and always
/// matches the history the child shows. `null`/empty when nothing is carried.
///
/// `snapshot_max_seq` is the highest `session_records.seq` the frontend saw when
/// it built the digest; the copy is capped at it so a sync that appends to the
/// parent between the two reads can't seed the child with turns the brief omitted.
#[tauri::command]
pub async fn fork_agent(
    supervisor: State<'_, Arc<Supervisor>>,
    ctx: State<'_, Arc<EngineCtx>>,
    parent_id: String,
    code: crate::supervisor::ForkCode,
    context: crate::supervisor::ForkContext,
    context_digest: Option<String>,
    snapshot_max_seq: Option<i64>,
) -> Result<AgentRecord> {
    let sup = supervisor.inner().clone();
    sup.fork_agent(
        ctx.inner().clone(),
        &parent_id,
        code,
        context,
        context_digest,
        snapshot_max_seq,
    )
    .await
}

#[tauri::command]
pub fn write_to_agent(
    supervisor: State<'_, Arc<Supervisor>>,
    ctx: State<'_, Arc<EngineCtx>>,
    agent_id: String,
    data: String,
) -> Result<()> {
    let sup = supervisor.inner().clone();
    sup.write_to_agent(ctx.inner(), &agent_id, data.as_bytes())
}

/// Returns `true` when the follow-up was enqueued for a later turn boundary
/// rather than delivered now (see `Supervisor::send_user_message`).
#[tauri::command]
pub fn send_user_message(
    supervisor: State<'_, Arc<Supervisor>>,
    ctx: State<'_, Arc<EngineCtx>>,
    agent_id: String,
    turn_id: String,
    text: String,
    attachments: Vec<String>,
) -> Result<bool> {
    let sup = supervisor.inner().clone();
    sup.send_user_message(ctx.inner(), &agent_id, &turn_id, &text, &attachments)
}

#[tauri::command]
pub fn answer_tool_use(
    supervisor: State<'_, Arc<Supervisor>>,
    agent_id: String,
    request_id: String,
    updated_input: serde_json::Value,
    behavior: ToolUseBehavior,
    message: Option<String>,
) -> Result<()> {
    supervisor
        .inner()
        .answer_tool_use(&agent_id, &request_id, updated_input, behavior, message)
}

#[tauri::command]
pub fn resize_agent(
    supervisor: State<'_, Arc<Supervisor>>,
    agent_id: String,
    cols: u16,
    rows: u16,
) -> Result<()> {
    supervisor.resize_agent(&agent_id, cols, rows)
}

#[tauri::command]
pub async fn resume_agent(
    supervisor: State<'_, Arc<Supervisor>>,
    ctx: State<'_, Arc<EngineCtx>>,
    agent_id: String,
) -> Result<()> {
    let sup = supervisor.inner().clone();
    sup.resume_agent(ctx.inner().clone(), &agent_id).await
}

#[tauri::command]
pub async fn switch_view(
    supervisor: State<'_, Arc<Supervisor>>,
    ctx: State<'_, Arc<EngineCtx>>,
    agent_id: String,
    view: AgentView,
) -> Result<()> {
    let sup = supervisor.inner().clone();
    sup.switch_view(ctx.inner().clone(), &agent_id, view).await
}

#[tauri::command]
pub async fn stop_agent(
    supervisor: State<'_, Arc<Supervisor>>,
    ctx: State<'_, Arc<EngineCtx>>,
    agent_id: String,
) -> Result<()> {
    let sup = supervisor.inner().clone();
    sup.stop_agent(ctx.inner().clone(), &agent_id).await
}

#[tauri::command]
pub async fn set_agent_effort(
    supervisor: State<'_, Arc<Supervisor>>,
    ctx: State<'_, Arc<EngineCtx>>,
    agent_id: String,
    effort: Option<String>,
) -> Result<()> {
    let sup = supervisor.inner().clone();
    sup.set_agent_effort(ctx.inner(), &agent_id, effort.as_deref())
        .await
}

#[tauri::command]
pub async fn set_agent_model(
    supervisor: State<'_, Arc<Supervisor>>,
    ctx: State<'_, Arc<EngineCtx>>,
    agent_id: String,
    model: Option<String>,
) -> Result<()> {
    let sup = supervisor.inner().clone();
    sup.set_agent_model(ctx.inner(), &agent_id, model.as_deref())
        .await
}

#[tauri::command]
pub async fn discard_agent(supervisor: State<'_, Arc<Supervisor>>, agent_id: String) -> Result<()> {
    let sup = supervisor.inner().clone();
    sup.discard_agent(&agent_id).await
}

#[tauri::command]
pub async fn archive_agent(
    supervisor: State<'_, Arc<Supervisor>>,
    ctx: State<'_, Arc<EngineCtx>>,
    agent_id: String,
) -> Result<()> {
    let sup = supervisor.inner().clone();
    sup.archive_agent(ctx.inner().clone(), &agent_id).await
}

#[tauri::command]
pub async fn restore_agent(
    supervisor: State<'_, Arc<Supervisor>>,
    ctx: State<'_, Arc<EngineCtx>>,
    agent_id: String,
) -> Result<()> {
    let sup = supervisor.inner().clone();
    sup.restore_agent(ctx.inner().clone(), &agent_id).await
}

#[tauri::command]
pub async fn add_repo_to_agent(
    supervisor: State<'_, Arc<Supervisor>>,
    ctx: State<'_, Arc<EngineCtx>>,
    agent_id: String,
    repo_path: String,
) -> Result<TrackedRepo> {
    let sup = supervisor.inner().clone();
    sup.add_repo_to_agent(ctx.inner().clone(), &agent_id, PathBuf::from(repo_path))
        .await
}
